File: examples/python/wavemq_binary_demo.py
#!/usr/bin/env python3
"""Thin demo wrapper around mbctl."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import time
from pathlib import Path


class MBctlError(RuntimeError):
    pass


def default_mbctl_path() -> Path:
    repo_root = Path(__file__).resolve().parents[2]
    return repo_root / "examples" / "bin" / "mbctl.exe"


def ensure_binary(path: Path) -> None:
    if path.is_file():
        return
    raise MBctlError(
        f"mbctl not found at {path}. Build it first with "
        r"powershell -ExecutionPolicy Bypass -File .\examples\build-mbctl.ps1"
    )


def run_mbctl(binary: Path, broker: str, args: list[str]) -> dict[str, object]:
    command = [str(binary), *args, "-broker", broker, "-json"]
    completed = subprocess.run(command, capture_output=True, text=True, check=False)

    if completed.returncode != 0:
        message = completed.stderr.strip() or completed.stdout.strip() or f"mbctl exit code {completed.returncode}"
        raise MBctlError(message)

    payload = completed.stdout.strip()
    if not payload:
        raise MBctlError(f"mbctl returned empty stdout for {' '.join(args)}")

    try:
        return json.loads(payload)
    except json.JSONDecodeError as exc:
        raise MBctlError(f"invalid JSON from mbctl for {' '.join(args)}: {exc}") from exc


def print_metadata(result: dict[str, object]) -> None:
    partitions = list(result.get("partitions", []))
    if not partitions:
        print("metadata: no partitions returned")
        return
    for item in partitions:
        print(
            "topic={topic} partition={partition} broker={broker} role={role} leader={leader} "
            "start={start} hwm={hwm} replicas={replicas} isr={isr}".format(
                topic=item["topic"],
                partition=item["partition"],
                broker=item["brokerId"],
                role=item["role"],
                leader=item["leader"],
                start=item["startOffset"],
                hwm=item["highWatermark"],
                replicas=item["replicas"],
                isr=item["isr"],
            )
        )


def print_fetch(result: dict[str, object]) -> None:
    records = list(result.get("records", []))
    print(f"high_watermark={result['highWatermark']} records={len(records)}")
    for record in records:
        print(
            "offset={offset} key={key} value={value}".format(
                offset=record["offset"],
                key=record["key"],
                value=record["value"],
            )
        )


def run_demo(args: argparse.Namespace) -> int:
    mbctl = args.mbctl.resolve()
    ensure_binary(mbctl)

    topic = args.topic or f"demo.binary.{int(time.time())}"
    group = args.group or f"demo-group-{int(time.time())}"

    print(f"broker={args.broker}")
    print(f"mbctl={mbctl}")

    print("1. ping")
    run_mbctl(mbctl, args.broker, ["ping"])
    print("pong")

    print(f"2. create topic {topic}")
    create_result = run_mbctl(
        mbctl,
        args.broker,
        [
            "create-topic",
            "-topic",
            topic,
            "-partitions",
            str(args.partitions),
            "-replication-factor",
            str(args.replication_factor),
        ],
    )
    print(
        "topic created partitions={partitions} rf={rf}".format(
            partitions=create_result["partitions"],
            rf=create_result["replicationFactor"],
        )
    )

    print("3. metadata")
    print_metadata(run_mbctl(mbctl, args.broker, ["metadata", "-topic", topic]))

    print("4. produce")
    base_offset = None
    for idx in range(args.count):
        produce_result = run_mbctl(
            mbctl,
            args.broker,
            [
                "produce",
                "-topic",
                topic,
                "-partition",
                str(args.partition),
                "-key",
                args.key,
                "-value",
                f"demo-message-{idx}",
            ],
        )
        if base_offset is None:
            base_offset = produce_result["baseOffset"]
    print(f"base_offset={base_offset}")

    print("5. list offsets")
    offsets = run_mbctl(
        mbctl,
        args.broker,
        ["list-offsets", "-topic", topic, "-partition", str(args.partition)],
    )
    print(f"earliest={offsets['earliest']} latest={offsets['latest']}")

    print("6. fetch")
    fetch_result = run_mbctl(
        mbctl,
        args.broker,
        [
            "fetch",
            "-topic",
            topic,
            "-partition",
            str(args.partition),
            "-offset",
            str(args.offset),
            "-max-bytes",
            str(args.max_bytes),
        ],
    )
    print_fetch(fetch_result)

    records = list(fetch_result.get("records", []))
    if records:
        committed_offset = int(records[-1]["offset"])
        print(f"7. commit offset group={group} offset={committed_offset}")
        run_mbctl(
            mbctl,
            args.broker,
            [
                "commit-offset",
                "-group",
                group,
                "-topic",
                topic,
                "-partition",
                str(args.partition),
                "-offset",
                str(committed_offset),
            ],
        )
        print("commit ok")

        print("8. fetch committed")
        committed = run_mbctl(
            mbctl,
            args.broker,
            [
                "fetch-committed",
                "-group",
                group,
                "-topic",
                topic,
                "-partition",
                str(args.partition),
            ],
        )
        print(f"committed offset={committed['offset']}")

    return 0


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="wave-mq demo wrapper built on mbctl")
    parser.add_argument("command", nargs="?", default="demo", choices=["demo"], help="demo scenario")
    parser.add_argument("--mbctl", type=Path, default=default_mbctl_path(), help="path to mbctl executable")
    parser.add_argument("--broker", default="127.0.0.1:7912", help="broker binary address")
    parser.add_argument("--topic", default="", help="topic name; default is generated")
    parser.add_argument("--group", default="", help="consumer group; default is generated")
    parser.add_argument("--partitions", type=int, default=1, help="topic partition count for demo")
    parser.add_argument("--replication-factor", type=int, default=1, help="topic replication factor for demo")
    parser.add_argument("--partition", type=int, default=0, help="partition used in demo")
    parser.add_argument("--count", type=int, default=3, help="messages to produce in demo")
    parser.add_argument("--offset", type=int, default=0, help="fetch start offset for demo")
    parser.add_argument("--max-bytes", type=int, default=1 << 20, help="fetch byte limit for demo")
    parser.add_argument("--key", default="", help="record key for produced messages")
    return parser


def main() -> int:
    args = build_parser().parse_args()
    try:
        return run_demo(args)
    except (OSError, MBctlError) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/matlab/mbctl/wavemq_binary_demo.m
% wave-mq demo wrapper built on mbctl.
% Build mbctl first with:
% powershell -ExecutionPolicy Bypass -File .\examples\build-mbctl.ps1

brokerAddr = "127.0.0.1:7912";
topic = sprintf("demo.binary.%d", floor(posixtime(datetime("now", "TimeZone", "UTC"))));
group = sprintf("demo-group-%d", floor(posixtime(datetime("now", "TimeZone", "UTC"))));
partition = int32(0);
values = [
    "hello from matlab"
    "second record"
    "third record"
];

scriptPath = mfilename("fullpath");
scriptDir = fileparts(scriptPath);
repoRoot = fileparts(fileparts(fileparts(scriptDir)));
mbctlPath = fullfile(repoRoot, "examples", "bin", "mbctl.exe");

if ~isfile(mbctlPath)
    error("wave:mbctl", "mbctl not found at %s. Run examples/build-mbctl.ps1 first.", mbctlPath);
end

fprintf("broker=%s\n", brokerAddr);
fprintf("mbctl=%s\n", mbctlPath);

fprintf("1. ping\n");
runMbctl(mbctlPath, brokerAddr, ["ping"]);
fprintf("pong\n");

fprintf("2. create topic %s\n", topic);
createResult = runMbctl(mbctlPath, brokerAddr, [
    "create-topic", ...
    "-topic", topic, ...
    "-partitions", "1", ...
    "-replication-factor", "1" ...
]);
fprintf("topic created partitions=%d rf=%d\n", createResult.partitions, createResult.replicationFactor);

fprintf("3. metadata\n");
metadataResult = runMbctl(mbctlPath, brokerAddr, ["metadata", "-topic", topic]);
printMetadata(metadataResult);

fprintf("4. produce\n");
baseOffset = [];
for idx = 1:numel(values)
    produceResult = runMbctl(mbctlPath, brokerAddr, [
        "produce", ...
        "-topic", topic, ...
        "-partition", string(partition), ...
        "-value", values(idx) ...
    ]);
    if isempty(baseOffset)
        baseOffset = produceResult.baseOffset;
    end
end
fprintf("base_offset=%d\n", baseOffset);

fprintf("5. list offsets\n");
offsetsResult = runMbctl(mbctlPath, brokerAddr, [
    "list-offsets", ...
    "-topic", topic, ...
    "-partition", string(partition) ...
]);
fprintf("earliest=%d latest=%d\n", offsetsResult.earliest, offsetsResult.latest);

fprintf("6. fetch\n");
fetchResult = runMbctl(mbctlPath, brokerAddr, [
    "fetch", ...
    "-topic", topic, ...
    "-partition", string(partition), ...
    "-offset", "0", ...
    "-max-bytes", "1048576" ...
]);
printFetch(fetchResult);

if isfield(fetchResult, "records") && ~isempty(fetchResult.records)
    records = fetchResult.records;
    committedOffset = records(end).offset;

    fprintf("7. commit offset group=%s offset=%d\n", group, committedOffset);
    runMbctl(mbctlPath, brokerAddr, [
        "commit-offset", ...
        "-group", group, ...
        "-topic", topic, ...
        "-partition", string(partition), ...
        "-offset", string(committedOffset) ...
    ]);
    fprintf("commit ok\n");

    fprintf("8. fetch committed\n");
    committedResult = runMbctl(mbctlPath, brokerAddr, [
        "fetch-committed", ...
        "-group", group, ...
        "-topic", topic, ...
        "-partition", string(partition) ...
    ]);
    fprintf("committed offset=%d\n", committedResult.offset);
end


function result = runMbctl(mbctlPath, brokerAddr, args)
cmdArgs = [ ...
    string(mbctlPath)
    args(:).'
    "-broker", brokerAddr
    "-json"
];

quoted = strings(1, numel(cmdArgs));
for idx = 1:numel(cmdArgs)
    quoted(idx) = quoteArg(cmdArgs(idx));
end

command = strjoin(quoted, " ");
[status, output] = system(char(command));
if status ~= 0
    error("wave:mbctl", "mbctl failed: %s", strtrim(output));
end

payload = strtrim(output);
if strlength(string(payload)) == 0
    error("wave:mbctl", "mbctl returned empty stdout");
end

result = jsondecode(payload);
if ~isfield(result, "ok") || ~result.ok
    error("wave:mbctl", "unexpected mbctl response");
end
end


function printMetadata(result)
if ~isfield(result, "partitions") || isempty(result.partitions)
    fprintf("metadata: no partitions returned\n");
    return
end

for idx = 1:numel(result.partitions)
    item = result.partitions(idx);
    fprintf( ...
        "topic=%s partition=%d broker=%d role=%s leader=%d start=%d hwm=%d replicas=%s isr=%s\n", ...
        item.topic, ...
        item.partition, ...
        item.brokerId, ...
        item.role, ...
        item.leader, ...
        item.startOffset, ...
        item.highWatermark, ...
        mat2str(item.replicas), ...
        mat2str(item.isr));
end
end


function printFetch(result)
if ~isfield(result, "records")
    fprintf("high_watermark=%d records=0\n", result.highWatermark);
    return
end

records = result.records;
fprintf("high_watermark=%d records=%d\n", result.highWatermark, numel(records));
for idx = 1:numel(records)
    record = records(idx);
    fprintf("offset=%d key=%s value=%s\n", record.offset, record.key, record.value);
end
end


function value = quoteArg(arg)
value = '"' + string(arg) + '"';
end
